grpc/server.go

```diff
@@ -169,17 +194,81 @@ func (sb *serverBuilder) Build(tlsConfig *tls.Config, statsRegistry prometheus.R
 		return server.Serve(listener)
 	}
 
+	// Initialize long-running health checks.
+	healthCtx, stopHealthChecks := context.WithCancel(context.Background())
+	for _, s := range sb.services {
+		check, ok := s.impl.(checker)
+		if !ok {
+			continue
+		}
+		sb.initLongRunningCheck(healthCtx, s.desc.ServiceName, 0, check, logger)
+	}
+
 	// Start a goroutine which listens for a termination signal, and then
 	// gracefully stops the gRPC server. This in turn causes the start() function
 	// to exit, allowing its caller (generally a main() function) to exit.
 	go cmd.CatchSignals(func() {
-		healthSrv.Shutdown()
+		stopHealthChecks()
+		sb.healthSrv.Shutdown()
 		server.GracefulStop()
 	})
 
 	return start, nil
 }
 
+func (sb *serverBuilder) initLongRunningCheck(healthCtx context.Context, service string, interval time.Duration, check checker, log blog.Logger) {
+	if interval <= 0 {
+		interval = 5 * time.Second
+	}
+
+	go func() {
+		ticker := time.NewTicker(interval)
+		defer ticker.Stop()
+
+		// Set the initial health status for the service.
+		sb.healthSrv.SetServingStatus(service, healthpb.HealthCheckResponse_NOT_SERVING)
```

Comment:
This looks good to me. Initially I thought that maybe `HealthCheckResponse_UNKNOWN` would be a better first state. However after checking [gRPC health check docs](https://github.com/grpc/grpc/blob/master/doc/health-checking.md), this state makes sense. Docs state:
> The server should register all the services manually and set the individual status, including an empty service name and its status. For each request received, if the service name can be found in the registry, a response must be sent back with an OK status and the status field should be set to SERVING or NOT_SERVING accordingly. If the service name is not registered, the server returns a NOT_FOUND GRPC status.

The service name is found in the registry because `Add` and `AddWithCheckInterval` register it there.

Reply:
> The service name is found in the registry because Add and AddWithCheckInterval register it there.

This is almost 100% correct. The service name is not found in the health service registry until it is registered by calling `*health.Server.SetServingStatus()`, the comment for which states:

> SetServingStatus is called when need to reset the serving status of a service or insert a new service entry into the statusMap.

So, the first time that we call this method for say 'sa.StorageAuthorityReadOnly', which happens inside of `grpc.serverBuilder.initLongRunningCheck()`, is also the first time that the health server's service registry becomes aware that 'sa.StorageAuthorityReadOnly' is even a service that it should be able to serve the health status for. If this call doesn't take place then 'sa.StorageAuthorityReadOnly' won't be included in the health service mapping but it will still function as a gRPC service on this gRPC server. In fact this is and will be the state of any component that doesn't implement the `grpc.checker` interface. Adding a service to the health service mapping just means that messages sent to the health service requesting the status of 'sa.StorageAuthorityReadOnly' will receive the registered health status instead a gRPC error with a status code of 'Not Found':

https://github.com/letsencrypt/boulder/blob/f47fcf576a9c2dce83642b5ccadca161c6a64698/vendor/google.golang.org/grpc/health/server.go#L62